### PR TITLE
Validate token trigger workflow

### PR DESCRIPTION
**Pull request by danidoni**
including
- definition of the new Token:Workflow,
- validation of the incoming webhook (check if it's a PR/MR with the right actions),
- extraction of some of the data we need from the payload (we still need to add those needed to report back)

### CODE REVIEW

**Comment by eduardoj** [comment]
[Create Token::Workflow](https://trello.com/c/trXgK4jU/1644-create-tokenworkflow)

**Comment by dmarcoux** [comment]
Code which is only for `Token::Workflow` should go in that class or in an object which `Token::Workflow` calls. Right now, that code is in `TriggerController` or called by `TriggerController` (like `TriggerControllerService::GithubEventExtractor`).

It would be cleaner to pass the webhook payload to `Token::Workflow` and deal with it in that class.

File: src/api/app/services/trigger_controller_service/github_event_extractor.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    def valid?
+      @http_request.env['HTTP_X_GITHUB_EVENT'] &&
+        @http_request.env['HTTP_X_GITHUB_EVENT'] == 'pull_request' &&
+        @payload['action'].in?(ALLOWED_GITHUB_ACTIONS)
+    end
+
+    # TODO: We also have to deal with wwwurlencoded parameters
+    # TODO: What happens when some of the keys are missing?
+    def add_payload_attributes_to_token(token)
```
Comment (dmarcoux) [inline src/api/app/services/trigger_controller_service/github_event_extractor.rb:18]: The payload attributes should be stored in the EventSubscriptions, not in the token. The token might have multiple EventSubscriptions. So if we pass the webhook payload to the `Token::Workflow` call method in the `create` action of the `TriggerController`, then we don't need to store that payload in the token. In the end, we'll do something like this to create the EventSubscriptions (see the payload there with `options[:something]`):

```ruby
    EventSubscription.create!(eventtype: 'Event::BuildSuccess',
                              receiver_role: 'watcher',
                              user: user,
                              channel: 'scm',
                              enabled: true,
                              token: self,
                              payload: {
                                scm: :github,
                                repository_owner: options[:repository_owner],
                                repository_name: options[:repository_name],
                                commit_sha: options[:commit_sha]
                              })
```

File: src/api/app/policies/token/workflow_policy.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class Token
+  class WorkflowPolicy < ApplicationPolicy
+    def initialize(_user, record)
+      super(record.user, record)
+    end
+
+    def create?
+      return false unless record.user.is_active?
+
+      PackagePolicy.new(record.user, record.package_from_association_or_params).update?
```
Comment (hennevogel) [inline src/api/app/policies/token/workflow_policy.rb:10]: Which package is that?

Comment (eduardoj) [reply src/api/app/policies/token/workflow_policy.rb]: Line removed.

**Comment by danidoni** [comment]
Closing this in favor of #11060

**Pull request closed without merging**